sycl/test/reduction/reduction_ctor.cpp

```diff
@@ -0,0 +1,170 @@
+// and the combine() method of the aux class 'reducer'.
+
+#include <CL/sycl.hpp>
+#include <cassert>
+
+using namespace cl::sycl;
+
+template <typename T, class BinaryOperation, int N>
+struct init_data_t {
+  void initInputData(T IdentityVal,
+                     buffer<T, 1> &InBuf,
+                     T &ExpectedReduValue) {
+    ExpectedReduValue = IdentityVal;
+    BinaryOperation Op;
+    auto In = InBuf.template get_access<access::mode::write>();
+    for (int I = 0; I < N; ++I) {
+      In[I] = ((I + 1) % 5) + 1;
+      ExpectedReduValue = Op(ExpectedReduValue, In[I]);
+    }
+  }
+};
+
+template <typename T, int N>
+struct init_data_t<T, std::multiplies<T>, N> {
+  void initInputData(T IdentityVal, buffer<T, 1> &InBuf, T &ExpectedReduValue) {
+    ExpectedReduValue = IdentityVal;
+    std::multiplies<T> Op;
+    auto In = InBuf.template get_access<access::mode::write>();
+    for (int I = 0; I < N; ++I) {
+      In[I] = 1 + (((I % 37) == 0) ? 1 : 0);
+      ExpectedReduValue = Op(ExpectedReduValue, In[I]);
+    }
+  }
+};
+
+template <typename T, typename Reduction>
+void test_reducer(Reduction &Redu, T A, T B) {
+  typename Reduction::reducer_type Reducer;
+  Reducer.combine(A);
+  Reducer.combine(B);
+
+  typename Reduction::binary_operation BOp;
+  T ExpectedValue = BOp(A, B);
+  assert(ExpectedValue == Reducer.MValue &&
+         "Wrong result of binary operation.");
+}
+
+template <typename T, typename Reduction>
+void test_reducer(Reduction &Redu, T Identity, T A, T B) {
+  typename Reduction::reducer_type Reducer(Identity);
+  Reducer.combine(A);
+  Reducer.combine(B);
+
+  typename Reduction::binary_operation BOp;
+  T ExpectedValue = BOp(A, B);
+  assert(ExpectedValue == Reducer.MValue &&
+         "Wrong result of binary operation.");
+}
+
+template <typename T, int Dim, class BinaryOperation>
+class Known;
+template <typename T, int Dim, class BinaryOperation>
+class Unknown;
+
+template <typename T>
+struct Point {
+  Point() : X(0), Y(0) {}
+  Point(T X, T Y) : X(X), Y(Y) {}
+  Point(T V) : X(V), Y(V) {}
+  bool operator==(const Point &P) const {
+    return P.X == X && P.Y == Y;
+  }
+  T X;
+  T Y;
+};
+
+template <typename T>
+bool operator==(const Point<T> &A, const Point<T> &B) {
+  return A.X == B.X && A.Y == B.Y;
+}
+
+template <class T>
+struct PointPlus {
+  using P = Point<T>;
+  P operator()(const P &A, const P &B) const {
+    return P(A.X + B.X, A.Y + B.Y);
+  }
+};
+
+template <typename T, int Dim, class BinaryOperation>
+void testKnown(T Identity, T A, T B) {
+
+  BinaryOperation BOp;
+  buffer<T, 1> ReduBuf(1);
+
+  queue Q;
+  Q.submit([&](handler &CGH) {
+    accessor<T, Dim, access::mode::discard_write, access::target::global_buffer>
+        ReduAcc(ReduBuf, CGH);
+    auto Redu = intel::reduction(ReduAcc, BOp);
+    assert(Redu.getIdentity() == Identity &&
+           "Failed getIdentity() check().");
+    test_reducer(Redu, A, B);
+    test_reducer(Redu, Identity, A, B);
```

Comment:
Reading/writing from/to a device accessor should be illegal outside of kernel.
Is it possible to move this code to the kernel?

Reply:
Wait, this code does not read from the accessor. This code creates 'reducer' on host and reads on host. 'reducer' does not depend on CGH and device.

Reply:
Redu.Identity is also a host thing.

I'll create better tests for device when parallel_for is ready for review.

Reply:
Ok, please, add a comment that accessor is not used for reading and writing.

Reply:
Done. Thank you!